Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift

```diff
@@ -533,3 +606,103 @@ open class ChatMessageListVC:
     }
 }
+
+// MARK: - Backwards Compatibility DataSource Diffing
+
+@available(iOS 13.0, *)
+internal extension ChatMessageListVC {
+    /// Setup the `UITableViewDiffableDataSource`.
+    func setupDiffableDataSource(for listView: ChatMessageListView) {
+        let diffableDataSource = UITableViewDiffableDataSource<Int, ChatMessage>(
+            tableView: listView
+        ) { [weak self] _, indexPath, _ -> UITableViewCell? in
+            /// Re-use old `cellForRowAt` to maintain customer's customisations.
+            let cell = self?.tableView(listView, cellForRowAt: indexPath)
+            return cell
+        }
+
+        self.diffableDataSource = diffableDataSource
+        listView.dataSource = diffableDataSource
+
+        /// Populate the Initial messages data.
+        var snapshot = NSDiffableDataSourceSnapshot<Int, ChatMessage>()
+        snapshot.appendSections([0])
+        snapshot.appendItems(dataSource?.messages ?? [], toSection: 0)
+        diffableDataSource.apply(snapshot, animatingDifferences: false)
+    }
+
+    /// Transforms an array of changes to a diffable data source snapshot.
+    func updateMessagesSnapshot(with changes: [ListChange<ChatMessage>], completion: (() -> Void)?) {
+        var snapshot = diffableDataSource?.snapshot() ?? NSDiffableDataSourceSnapshot<Int, ChatMessage>()
+
+        var updatedMessages: [ChatMessage] = []
+        var removedMessages: [ChatMessage] = []
+        var insertedMessages: [(ChatMessage, row: Int)] = []
+        var movedMessages: [(from: ChatMessage, to: ChatMessage)] = []
+
+        var hasNewInsertions = false
+
+        changes.forEach { change in
+            switch change {
+            case let .insert(message, indexPath):
+                insertedMessages.append((message, row: indexPath.row))
+                hasNewInsertions = indexPath.row == 0
+            case let .update(message, _):
+                updatedMessages.append(message)
+            case let .remove(message, _):
+                removedMessages.append(message)
+            case let .move(_, fromIndex, toIndex):
+                guard let fromMessage = snapshot.itemIdentifiers[safe: fromIndex.row] else { break }
+                guard let toMessage = snapshot.itemIdentifiers[safe: toIndex.row] else { break }
+                movedMessages.append((from: fromMessage, to: toMessage))
+            }
+        }
+
+        let sortedInsertedMessages = insertedMessages
+            .sorted(by: { $0.row < $1.row })
+            .map(\.0)
+
+        if hasNewInsertions, let currentFirstMessage = snapshot.itemIdentifiers.first {
+            // Insert new messages at the bottom.
+            snapshot.insertItems(sortedInsertedMessages, beforeItem: currentFirstMessage)
+        } else if let currentLastMessage = snapshot.itemIdentifiers.last {
+            // Load new messages at the top.
+            snapshot.insertItems(sortedInsertedMessages, afterItem: currentLastMessage)
+        }
```

Comment:
This may be not the case but I have a suspicion that if the channel is empty, message insertions will never happen because of `snapshot.itemIdentifiers` being empty. Would it be possible to check it?

Reply:
Fixed ✅